From: Rene Kita <mail@rkta.de>
Subject: Reduce scope of dangerous macro

The Strcatc macro does no bounds checking when adding to the string.
This may lead to a out-of-bounds write violation.

It is only used in base64_encode where the needed size is known
beforehand. In this case it's ok (and cheaper) to omit the check.

Move the macro to the inner scope of the function to avoid programmers
using it.

While at it, do the same with the Strnulterm as it is only used once in
this function.

---
 Str.h |    2 --
 etc.c |    4 ++++
 2 files changed, 4 insertions(+), 2 deletions(-)

--- a/Str.h
+++ b/Str.h
@@ -69,8 +69,6 @@ void Strgrow(Str s);
 
 #define STR_SIZE_MAX (INT_MAX / 32)
 #define Strcat_char(x,y) (((x)->length+1>=STR_SIZE_MAX)?0:(((x)->length+1>=(x)->area_size)?Strgrow(x),0:0,(x)->ptr[(x)->length++]=(y),(x)->ptr[(x)->length]=0))
-#define Strcatc(x,y) ((x)->ptr[(x)->length++]=(y))
-#define Strnulterm(x) ((x)->ptr[(x)->length]=0)
 #define Strcmp(x,y)                  strcmp((x)->ptr,(y)->ptr)
 #define Strcmp_charp(x,y)            strcmp((x)->ptr,(y))
 #define Strncmp(x,y,n)               strncmp((x)->ptr,(y)->ptr,(n))
--- a/etc.c
+++ b/etc.c
@@ -2114,6 +2114,8 @@ static char Base64Table[] =
 Str
 base64_encode(const char *src, size_t len)
 {
+#define Strcatc(x,y) ((x)->ptr[(x)->length++]=(y))
+#define Strnulterm(x) ((x)->ptr[(x)->length]=0)
     Str dest;
     const unsigned char *in, *endw, *s;
     unsigned long j;
@@ -2170,4 +2172,6 @@ base64_encode(const char *src, size_t le
     }
     Strnulterm(dest);
     return dest;
+#undef Strcatc
+#undef Strnulterm
 }
